### PR TITLE
Provide a JavaScript library which gives you a Web3 object by default…

### PR DESCRIPTION
**Pull request by amateescu**
… and change the default server to the local Ganache dev node.

[Asset Packagist](http://asset-packagist.org) is currently not very reliable as a source for the Web3.js package, so I think we should provide a fixed version of the library for now.

I also changed the default server in this commit because I think most people who are trying out this module will probably start on a localhost setup, and it seems that Ganache is most popular option these days.

### CODE REVIEW

File: ethereum.module [hunk 1]
```diff
@@ -45,3 +47,13 @@ function ethereum_field_formatter_info_alter(array &$info) {
...
 }
+
+/**
+ * Implements hook_js_settings_alter().
+ */
+function ethereum_js_settings_alter(&$settings, AttachedAssetsInterface $assets) {
+  // Provide the URL of the current server in drupalSettings to allow JavaScript
+  // code to rely on an instantiated Web3 object.
+  $current_server = \Drupal::config('ethereum.settings')->get('current_server');
+  $settings['ethereum']['currentServerHost'] = EthereumServer::load($current_server)->getUrl();
```
Comment (digitaldonkey) [inline ethereum.module]: We risk publishing private Server information. 
If we provide a public url to connect it might be a different from the backend.

**Comment by digitaldonkey** [comment]
I was thinking about the more regularly updated [bower package](https://asset-packagist.org/package/bower-asset/web3). On the long term I hope for Asset Packagist support as maintaining a web3 as full part of Ethereum version is too much responsibility. 
But now, we should just integrate it as a "TxSigner". 

Transaction signer config entity

The Idea of Transaction signer config entities be to create some config and library loading into a small sub/module which then would provide a TX singing mechanism. 

The currently provided [mascara_wrapper](https://github.com/digitaldonkey/mascara_wrapper) is considered to provide a kind of "dapp launcher".  

There are now also contracts, which might attachable JS settings on the todo list. 

My I basically want to allow a admin to decide on TX signers, but with Network validation etc i think i would be good to have some kind of drupalish app launcher standard providing a defined environment for starting Dapp's What you think?

File: ethereum.module (hunk 1 above)
Comment (amateescu) [reply ethereum.module]: Good point! I added a new 'frontend_server' setting so we can select different servers for backend / frontend.

**Comment by amateescu** [comment]
> I was thinking about the more regularly updated bower package.

This was my preferred option as well, but the 1.0 branch of web3.js has a dependency on bn.js (https://github.com/ethereum/web3.js/blob/1.0/bower.json#L11), which does not have a release on bower: https://bower.io/search/?q=bn , so it is also not available on Asset Packagist.

> On the long term I hope for Asset Packagist support as maintaining a web3 as full part of Ethereum version is too much responsibility.

I also don't really like the current solution, but it's the only one that works reliably. As for maintenance, we just have to update the web3.js version number from time to time, just like Drupal core does with its JS (and PHP) dependencies.

> But now, we should just integrate it as a "TxSigner".

IMO, having a Web3 object always available on the frontend allows our users to write any kind of interaction with an Ethereum node, and transaction signers, "dapp launchers" and everything else can build on top of it.

There's also the problem that the module's code on drupal.org can not include GPL v3 libraries in its repository, so we need to provide it as a composer dependency like proposed in this PR.

**Comment by digitaldonkey** [comment]
**GPL-v3**
I see was not aware of this GPL-v3 thing (Just for reference https://www.drupal.org/project/erpal/issues/2387307).

**asset-packagelist.org**
So maybe asset-packagelist.org is not the solution. I'm just afraid, we will end up forcing a user to add tons of library resources to toot composer.json. Are Drupal devs used to this now?

>web3.js has a dependency on bn.js 

Yes. but Bignumber is included in the web3 distribution.
https://github.com/ethereum/web3.js/blob/59aae306c1c31ef6a65b9196e7f03af74c69e059/dist/web3.js#L13613

**Frontend**

There are currently a lot of projects very similar to the "mascara_wrapper" in the space and I hope we will end up at some day with a solution, which supports the majority of TX-Signers available. 

That's the theory :roll_eyes:

I like your solution, because it's simple and provide a easy staring point, but I see two problems: 

With the code is that I basically use the same if/else when I try to add the Mascara plugin, so I don't think they can run in parallel. 

https://github.com/digitaldonkey/ethereum/blob/680f622dc80c597ac40990b9c91d4c785d0244e3/ethereum_txsigner/js/mascara/src/mascara/index.js#L54-L67

I want to keep to try frontend implementations very un-opinionated (e.g. there is also [ethjs](https://github.com/ethjs)). 

I would propose that we add this as a second TX-signer and let the user chose.  If we do it in the TX signer module we can also reuse the web3 library.

**Comment by digitaldonkey** [comment]
We have it in now, but we really need to review this approach of TX-signers in this context. 

What I'm also worried about is that we might easily expose a username/Password to random frontend users unintentionally if you use a password/token protected Frontend server.

Added a ticket https://github.com/digitaldonkey/ethereum/issues/29

**Pull request merged**